src/hip_memory.cpp

```diff
@@ -33,10 +33,63 @@ THE SOFTWARE.
 __device__ char __hip_device_heap[__HIP_SIZE_OF_HEAP];
 __device__ uint32_t __hip_device_page_flag[__HIP_NUM_PAGES];
 
+#define IMAGE_PITCH_ALIGNMENT 256
+
 // Internal HIP APIS:
 namespace hip_internal {
 
 namespace {
+
+    template <typename T> inline T alignDown(T value, size_t alignment) {
+      return (T)(value & ~(alignment - 1));
+    }
+
+    template <typename T> inline T* alignDown(T* value, size_t alignment) {
+      return (T*)alignDown((intptr_t)value, alignment);
+    }
+
+    template <typename T> inline T alignUp(T value, size_t alignment) {
+      return alignDown((T)(value + alignment - 1), alignment);
+    }
+
+    template <typename T> inline T* alignUp(T* value, size_t alignment) {
+      return (T*)alignDown((intptr_t)(value + alignment - 1), alignment);
+    }
+
+    size_t getNumChannels(hsa_ext_image_channel_order_t channelOrder) {
+      switch (channelOrder) {
+        case HSA_EXT_IMAGE_CHANNEL_ORDER_RG:
+          return 2;
+        case HSA_EXT_IMAGE_CHANNEL_ORDER_RGB:
+          return 3;
+        case HSA_EXT_IMAGE_CHANNEL_ORDER_RGBA:
+          return 4;
+        case HSA_EXT_IMAGE_CHANNEL_ORDER_R:
+        default:
+          return 1;
+      }
+    }
+
+    size_t getElementSize(hsa_ext_image_channel_order_t channelOrder, hsa_ext_image_channel_type_t channelType) {
+      size_t bytesPerPixel = getNumChannels(channelOrder);
+      switch (channelType) {
+        case HSA_EXT_IMAGE_CHANNEL_TYPE_UNSIGNED_INT8:
+        case HSA_EXT_IMAGE_CHANNEL_TYPE_SIGNED_INT8:
+          break;
+
+        case HSA_EXT_IMAGE_CHANNEL_TYPE_SIGNED_INT32:
+        case HSA_EXT_IMAGE_CHANNEL_TYPE_UNSIGNED_INT32:
+        case HSA_EXT_IMAGE_CHANNEL_TYPE_FLOAT:
+          bytesPerPixel *= 4;
+          break;
+
+        default:
+          bytesPerPixel *= 2;
+          break;
+      }
+      return bytesPerPixel;
+    }
+
```

Comment:
We could move this to hip_hcc_internal.h. This way we don't need to duplicate this in both hip_memory.cpp and hip_texture.cpp.
Also from the below code I only see alignUp(T value, size_t alignment) variant being used. Do we need the remaining?

Reply:
done.